src/G2/Equiv/Induction.hs

```diff
@@ -0,0 +1,345 @@
+      working_info = [(sc1, sc2, pm) | ((sc1, sc2), Just pm) <- mr_zipped]
+      working_info' = filter (\(_, _, PrevMatch _ (_, p2) _ p_outer2) -> validScrutinee s2 p2 p_outer2) working_info
+  case working_info' of
+    [] -> return Nothing
+    h:_ -> let (sc1, sc2, PrevMatch _ (p1, p2) (mapping, _) p_outer2) = h
+               e2_old = exprExtract p_outer2
+               mapping_list = HM.toList mapping
+               e2_old' = foldr (\(i, e) acc -> replaceASTs (Var i) e acc) e2_old mapping_list
+               e1_new = replaceScrutinee sc1 e2_old' $ exprExtract s1
+               h1_new = E.union (expr_env s1) (expr_env p_outer2)
+               s1' = s1 {
+                 curr_expr = CurrExpr Evaluate e1_new
+               , expr_env = h1_new
+               }
+               -- TODO should result include generalization?
+               -- TODO p1, p2, q1, q2, present scrutinees
+               -- placeholders for real present and name to avoid undefined errors
+               im = IndMarker {
+                 ind_real_present = (s1, s2)
+               , ind_used_present = (s1, s2)
+               , ind_past = (p1, p_outer2)
+               , ind_result = (s1', s2)
+               , ind_present_scrutinees = (sc1, sc2)
+               , ind_past_scrutinees = (p1, p2)
+               , ind_side = ILeft
+               , ind_fresh_name = Name "" Nothing 0 Nothing
+               }
+           in return $ Just (s1', im)
+
+-- This attempts to perform induction on both the left side and the right side.
+-- Precedence goes to the left side:  induction will only happen on the right
+-- if induction on the left fails.
+induction :: S.Solver solver =>
+             solver ->
+             HS.HashSet Name ->
+             [(StateET, StateET)] ->
+             (StateET, StateET) ->
+             W.WriterT [Marker] IO (Maybe (StateET, StateET, IndMarker))
+induction solver ns prev (s1, s2) = do
+  ind <- inductionL solver ns prev (s1, s2)
+  case ind of
+    Just (s1l, iml) -> return $ Just (s1l, s2, iml)
+    Nothing -> do
+      let prev' = map swap prev
+      ind' <- inductionL solver ns prev' (s2, s1)
+      case ind' of
+        Just (s2r, imr) -> return $ Just (s1, s2r, imr)
+        Nothing -> return Nothing
+
+-- TODO complex conditional, but avoids needless generalization
+-- TODO returning marker in case of failure
+-- This attempts to perform both left-side and right-side induction with
+-- multiple different combinations of present states.  The actual substitution
+-- for induction can happen on either side.  All of the present-state
+-- combinations tried have the same left-hand state, but the right-hand state
+-- varies between attempts.
+inductionFoldL :: S.Solver solver =>
+                  solver ->
+                  HS.HashSet Name ->
+                  Name ->
+                  (StateH, StateH) ->
+                  (StateET, StateET) ->
+                  W.WriterT [Marker] IO (Int, StateET, StateET, IndMarker)
```

Comment:
This looks like a case where we should be returning a Maybe instead of having an Int?  (Or, if we really need the Int for some reason, we should be returning a `Maybe (Int, StateET, StateET, IndMarker)` rather than giving a special meaning to a negative Int.)